Fix blocky appearance of water surfaces

niligulmohar opened the pull request:
The water surface pixel shader implements a kind of dithering effect for
its reflections and refractions that appears like blocks when the
resolution is increased. This can be seen for example in New Londo Ruins
or at the Darkroot Basin. The dithering effect is implemented by
multiplying viewport coordinates with parts of register c164. The
register is set each frame to contain half the expected resolution of
the render buffer in xy, and the reciprocals of the resolution in zw.

This change improves the appearance of the water surface shader by
intercepting updates to register c164 and setting its values to reflect
the actual resolution of the render buffer.

niligulmohar committed:
Fix blocky appearance of water surfaces

The water surface pixel shader implements a kind of dithering effect for
its reflections and refractions that appears like blocks when the
resolution is increased. This can be seen for example in New Londo Ruins
or at the Darkroot Basin. The dithering effect is implemented by
multiplying viewport coordinates with parts of register c164. The
register is set each frame to contain half the expected resolution of
the render buffer in xy, and the reciprocals of the resolution in zw.

This change improves the appearance of the water surface shader by
intercepting updates to register c164 and setting its values to reflect
the actual resolution of the render buffer.

SalamatiQus commented:
Playing at 1920x1080 I have noticed that this patch is doing wonders while DoF is set to 540. However, increasing DoF makes water appear blocky agian. Is there a way to fix this?

niligulmohar commented:
I see the problem. I suspect i guessed wrong about the values being just half the resolution. It might be that the values should be the resolution of the DoF buffer.

I'll see if that will fix it.

niligulmohar committed:
Update water surface blockiness fix

The water surface is rendered into a buffer that has the same resolution
as the first DoF buffer, and the values passed in c164 now reflect this.

The dithering effect is not actually used by the unmodified game, and
values passed through c164 should match the exact resolution of the
render buffer used for the water surface. This is also required if the
water surface is to look good in all choices of DoF buffer resolution.

niligulmohar commented:
When running without DSfix i realize that the dithering effect is not actually used. The dithering was incorrect, but looked ok on high enough resolutions. Also, the water surface is drawn into a render buffer that is the same resolution as the first DoF buffer instead of at the full resolution as i thought.

The latest commit fixes this by using the DoF buffer resolution for the shader instead.

SalamatiQus commented:
I have applied this change and now it looks fantastic!
Thank you very much!

dreambottle commented:
Great improvements! Hope this gets merged soon.

Keith94 commented:
@niligulmohar Will you release any custom builds of DSfix with your changes included?

niligulmohar commented:
@Keith94 Not at the moment, sorry.

I've reached out to PeterTh to see if i can.

Currently, license issues prevents making releases. (The GPL requires extra permissions from the copyright owner for linking with libraries with incompatible licenses, in this case MS Detours.)

theultramage commented:
Just fyi, MS Detours has at some point been re-licensed as MIT and is GPL-compatible.
